towerlib/towerlib.py

```diff
@@ -1997,24 +2011,579 @@ def schedules(self):
+                                    format(existing_job_type))
+                        else:
+                            self._logger.debug("Changing job type from '{}' to '{}'"
+                                               .format(existing_job_type, new_job_type))
+                            job_template_data = {
+                                "id": job_template.id,
+                                "job_type": new_job_type,
+                                "name": job_template.name,
+                                "playbook": job_template.playbook
+                            }
+                            self.change_job_template_data(job_template_data)
+            if count < 1:
+                self._logger.debug("No job template, which matched all the labels in the config file.")
+
+    def get_credential_id_from_existing_project_by_scm_url(self, scm_url):
+        """This function gets credential id from an existing project, which has the same credential id.
+
+        This function is for additional convenience. If there is already a project with the same scm url configured,
+        then we will find it with this function. This function also makes sure that the user does not have to specify a
+        credential id in the config file .
+        This function will only be called if no credential id is given in the config file. Anyway, if no credential is
+        given by config file and no credential id is found with this function, the script will continue and create
+        projects, because projects can be created in Ansible Tower without credentials (but in that case, the scm
+        update job will fail.)
+
+        Args:
+            scm_url: the scm_url of the corresponding project.
+
+        Returns:
+            id of a credential if corresponding project exists, else None.
+
+        """
+        scm_credential_type_ids = []
+        credential_types = list(self.credential_types)
+        for credential_type in credential_types:
+            if credential_type.kind == 'scm':
+                scm_credential_type_ids.append(credential_type.id)
+        credential_id = None
+        projects = list(self.projects)
+        for scm_credential_type_id in scm_credential_type_ids:
+            for project in projects:
+                if project.scm_url == scm_url:
+                    self._logger.debug("Project found with the scm credential.")
+                    if project.credential.credential_type.id == scm_credential_type_id:
+                        credential_id = project.credential.id
+                        if credential_id:
+                            self._logger.debug("Credential ID found from the existing project '{}' and the id is '{}'"
+                                               .format(project.name, credential_id))
+                            return credential_id
+        return credential_id
+
+    def get_projects_by_scm_url_and_branch(self, scm_url, scm_branch):
```

Comment:
I think that all of these can be implemented client side as a filter on the projects and should not be part of the library.